### PR TITLE
MRG: Add whitening to plots

### LINKED ISSUE #3926: evoked.plot_white() for raw data
An equivalent of evoked.plot_white() for raw data would be very useful for people who want to compare whitening methods for raw data (e.g. resting state) or empty room data.

### PR DESCRIPTION
**Pull request by larsoner**
Need to:

- [x] fill out docstrings
- [x] add tests
- [x] update whats_new.rst

Will post CircleCI result once it comes up.

Closes #3926.

### CODE REVIEW

**Comment by larsoner** [comment]
Certainly not the prettiest example I've ever made:

https://6600-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.AXJHm7i/html/auto_tutorials/plot_whitened.html

**Comment by larsoner** [comment]
cc @olafhauk feel free to try this out to see if it addresses your use case from #3926

**Comment by larsoner** [comment]
This is now ready for review/merge from my end

**Comment by agramfort** [comment]
see texts overlap

https://6600-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.AXJHm7i/html/_images/sphx_glr_plot_whitened_009.png

**Comment by agramfort** [comment]
if you pass a noise cov that does not contain a certain channel you should remove it as presently when I click on a bad channel I get errors such as:

```
  File "/Users/alex/work/src/mne-python/mne/cov.py", line 1639, in compute_whitener
    % missing)
RuntimeError: Not all channels present in noise covariance:
['MEG 2443']
```

**Comment by larsoner** [comment]
@agramfort fixed, now if you un-mark a bad it disappears from the plot, feel free to try it. I also fixed the overlapping name issue in the example, and a small bug with switching to/from `butterfly` mode.

Here are the updated examples:

https://6665-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.O3jInQw/html/auto_tutorials/plot_whitened.html

https://6665-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.O3jInQw/html/auto_tutorials/plot_visualize_raw.html

**Comment by agramfort** [comment]
@larsoner you need to rebase this one

File: mne/viz/epochs.py [hunk 1]
```diff
@@ -750,6 +749,12 @@ def plot_epochs(epochs, picks=None, scalings=None, n_epochs=20, n_channels=20,
         at least a 120 Hz displayed sample rate).
 
         .. versionadded:: 0.15
+    noise_cov : instance of Covariance | str | None
+        Noise covariance used to whiten the data while plotting.
+        Whitened data channels are scaled by ``scalings['whitened']``,
+        and their channel names are shown in italic.
```
Comment (agramfort) [inline mne/viz/epochs.py]: why the str option?

Comment (agramfort) [reply mne/viz/epochs.py]: now I see that it is to be able to read from file from disk.
Please document it.

File: mne/viz/tests/test_evoked.py [hunk 2]
```diff
@@ -78,181 +78,184 @@ def test_plot_evoked():
     import matplotlib.pyplot as plt
     rng = np.random.RandomState(0)
     evoked = _get_epochs().average()
```
Comment (agramfort) [inline mne/viz/tests/test_evoked.py]: can you explain me the big diff in this file?

**Comment by agramfort** [comment]
if I toggle off the bad eeg channel in the epochs.plot I get:

```
Traceback (most recent call last):
  File "/Users/alex/anaconda/envs/mne/lib/python3.6/site-packages/matplotlib/cbook/__init__.py", line 389, in process
    proxy(*args, **kwargs)
  File "/Users/alex/anaconda/envs/mne/lib/python3.6/site-packages/matplotlib/cbook/__init__.py", line 227, in __call__
    return mtd(*args, **kwargs)
  File "/Users/alex/work/src/mne-python/mne/viz/epochs.py", line 1490, in _mouse_click
    params['label_click_fun'](pos)
  File "/Users/alex/work/src/mne-python/mne/viz/epochs.py", line 1462, in _pick_bad_channels
    params['plot_update_proj_callback'](params)
  File "/Users/alex/work/src/mne-python/mne/viz/epochs.py", line 1342, in _plot_update_epochs_proj
    params['info'], params['noise_cov'], True, params['use_noise_cov'])
  File "/Users/alex/work/src/mne-python/mne/viz/utils.py", line 2269, in _setup_plot_projector
    noise_cov, use_info, verbose=False)
  File "<string>", line 2, in compute_whitener
  File "/Users/alex/work/src/mne-python/mne/utils.py", line 727, in verbose
    return function(*args, **kwargs)
  File "/Users/alex/work/src/mne-python/mne/cov.py", line 1640, in compute_whitener
    % missing)
RuntimeError: Not all channels present in noise covariance:
['EEG 053']
```

also when I do evoked.plot(noise_cov=noise_cov) and select a time interval the topos that are produced look weird.

File: mne/viz/tests/test_evoked.py [hunk 3]
```diff
@@ -78,181 +78,184 @@ def test_plot_evoked():
...
-        assert_raises(RuntimeError, evoked_delayed_ssp.plot,
-                      proj='interactive')
-        assert_raises(RuntimeError, evoked_delayed_ssp.plot,
-                      proj='interactive', axes='foo')
-        plt.close('all')
 
-        # test GFP only
-        evoked.plot(gfp='only')
-        assert_raises(ValueError, evoked.plot, gfp='foo')
+    evoked.plot(noise_cov=cov_fname)
```
Comment (larsoner) [inline mne/viz/tests/test_evoked.py]: this is the new line

File: mne/viz/tests/test_evoked.py (hunk 2 above)
Comment (larsoner) [reply mne/viz/tests/test_evoked.py]: There is no reason to have the `with warnings.catch_warnings` in there -- it could be masking useful warnings (and might have been, don't remember)

**Comment by larsoner** [comment]
> when I do evoked.plot(noise_cov=noise_cov) and select a time interval the topos that are produced look weird.

I think this likely isn't bad math, but rather just whitened topos looking weird.

**Comment by larsoner** [comment]
> if I toggle off the bad eeg channel in the epochs.plot I get [an error]

Actually I think this is reasonable behavior and error message. If you try to plot whitened data and your noise cov does not have all data channels, you should get an error, no?

Other comments addressed and rebased.

**Comment by agramfort** [comment]
I think we should not display signals that are not part of the noise
covariance. That would be the easiest.

**Comment by larsoner** [comment]
Okay lines have been removed

**Comment by larsoner** [comment]
CI failure is a sporadic unrelated Travis OSX problem

**Comment by dengemann** [comment]
@larsoner @agramfort why would we need evoked.plot(noise_cov=noise_cov) if there is evoked.plot_white? How should a user understand when to use which one? I am not sure what is the answer.

**Comment by larsoner** [comment]
The former is (consistent with raw and epochs options and) more useful for
browsing data and responses, the latter is more useful for diagnostic plots
for the covariance.

**Comment by larsoner** [comment]
Any chance for review here? I'd like to move on to doing `group_by` support in `epochs.plot` but don't want to start until this is in to avoid conflicts.

**Comment by agramfort** [comment]
just played a bit. Behavior with bad channels is still not super robust.

See

[image: Inline image 1]

Basically you see that the bad channel in whiten mode is not plotted but
the label is not in Italic. You can also click on the bad in whiten mode
and you
get a gray line although in whiten mode the bad channels should just be
not plotted at all.

**Comment by larsoner** [comment]
@agramfort done. (Also the image you tried to post did not post, but I think I understood what you meant: never show traces of bads from `info` or `cov` in whiten mode.)

**Comment by agramfort** [comment]
you should disable toggling bad channels when clicking on channel name in
whitened mode.

marking a good channel as bad in whitened mode gives:

Traceback (most recent call last):
  File
"/Users/alex/anaconda/envs/mne/lib/python3.6/site-packages/matplotlib/cbook/__init__.py",
line 389, in process
    proxy(*args, **kwargs)
  File
"/Users/alex/anaconda/envs/mne/lib/python3.6/site-packages/matplotlib/cbook/__init__.py",
line 227, in __call__
    return mtd(*args, **kwargs)
  File "/Users/alex/work/src/mne-python/mne/viz/epochs.py", line 1491, in
_mouse_click
    params['label_click_fun'](pos)
  File "/Users/alex/work/src/mne-python/mne/viz/epochs.py", line 1463, in
_pick_bad_channels
    params['plot_update_proj_callback'](params)
  File "/Users/alex/work/src/mne-python/mne/viz/epochs.py", line 1343, in
_plot_update_epochs_proj
    params['info'], params['noise_cov'], True, params['use_noise_cov'])
  File "/Users/alex/work/src/mne-python/mne/viz/utils.py", line 2278, in
_setup_plot_projector
    for pick in good_picks]
AssertionError

**Comment by larsoner** [comment]
I think it would we unexpected not to be able to toggle bad channels when
in whitening mode. It violates "least surprise" to me at least, and removes
functionality without any real reason to do so.

I'd rather restore the "bad as gray, not in covariance as transparent"
functionality that was there before (and fix the errors of course). Then
users can toggle bads regardless of mode.

**Comment by agramfort** [comment]
>
> I think it would we unexpected not to be able to toggle bad channels when
> in whitening mode. It violates "least surprise" to me at least, and removes
> functionality without any real reason to do so.
>
> I'd rather restore the "bad as gray, not in covariance as transparent"
> functionality that was there before (and fix the errors of course). Then
> users can toggle bads regardless of mode.


ok but you have 4 types of channels (bad and in cov, bad and not in cov,
good in cov, good not in cov)

we need to make sure it works for the 4 cases.

**Comment by larsoner** [comment]
I will treat "bad and in cov" as "not in cov", since in practice you don't usually reenable these. Then it's simpler, and the interactive experience should mirror what you could get by setting `info['bads']` in code (which is the same way it works when whitening is off).

**Comment by larsoner** [comment]
(sorry I mean treat "bad and in cov['bads']" as "not in cov")

**Comment by agramfort** [comment]
sounds like a plan

**Comment by dengemann** [comment]
I am -0.5 against the whitening arg in plot_evoked. I am afraid it dilutes
API. Users may get confused and miss that there is plot_evoked_white. What
is the motivation here?

**Comment by larsoner** [comment]
@agramfort done

@dengemann:

1. `plot_white` is more useful for diagnostic plots for the covariance. `plot_evoked` is more useful for browsing data and looking at responses (it is less crowded with extra cov-related "stuff").
2. More importantly, though, is consistency. It is an argument for `raw.plot`, `epochs.plot`, and `evoked.plot_topo`, so *not* having it as an argument for `evoked.plot` would be unexpected for users.

**Comment by larsoner** [comment]
@dengemann for you:

https://github.com/mne-tools/mne-python/pull/4799/commits/518ff57cea2a97ac8f9d58ae9061e455cd6f8845

**Comment by dengemann** [comment]
Let’s just understand a bit the overall concept here. Can you share some
insights on what this is enabling? General purpose / expert user
functionality for visualizing whitening? Curious to learn what is
motivating you.

**Comment by larsoner** [comment]
Read #3926

**Comment by agramfort** [comment]
xlabels on epochs viewer shows byte and not str on python 3 so it does not render fine:

screen shot 2018-02-03 at 11 25 52

**Comment by agramfort** [comment]
with:

```
evoked.plot(noise_cov=noise_cov)
evoked.plot_white(noise_cov=noise_cov)
```

the scales for evoked data do not match:

screen shot 2018-02-03 at 11 28 27

with plot_white we did the proper scalings so during baseline the white data have actually a std of approx 1. This is not the case with your plot. How did you scale this? also for Raw and Epochs?

**Comment by larsoner** [comment]
> How did you scale this? also for Raw and Epochs?


The same way for each, just by doing a dot product with the whitener.

I could look into doing the scaling the same way, but I don't like that it
departs from the "this is how your data are scaled for localization" idea /
equivalence. So maybe this should be a difference between plot and
plot_white with a proper documentation update. (Also if you whiten with the
additional plot_white scaling and an empty room covariance, the traces
might be huge in scale, which isn't as useful for data browsing.)

**Comment by larsoner** [comment]
(Actually it is probably also huge with this implementation, so never mind
that bit.)

**Comment by larsoner** [comment]
> xlabels on epochs viewer shows byte and not str on python 3 so it does not render fine:

... actually I'm not sure how my changes could have caused this as I just change the styling not the content of the lines. Does this happen on `master` for you? It looks like a `matplotlib` bug to me.

**Comment by larsoner** [comment]
I can confirm the `b''` behavior happens on `master` as well, so let's not address it here.

So the only issue that I can see is the scaling. Let me know if you think it needs to be the same, or if it's okay to have it be different.

**Comment by larsoner** [comment]
Okay scaling fixed / should now match that of `plot_white`. Along the way I refactored/simplified of some of the rank calculations.

I also fixed a bug with `plot_evoked_white`:
```
import mne
fname = mne.datasets.testing.data_path() + '/SSS/test_move_anon_raw_spatiotemporal_4s_sss.fif'
raw = mne.io.read_raw_fif(fname)
raw.set_eeg_reference(projection=True)
events = mne.make_fixed_length_events(raw, 1)
epochs = mne.Epochs(raw, events)
cov = mne.compute_covariance(epochs, tmax=0.)
evoked = epochs.average()
evoked.plot_white(cov)
```
on `master`:

master

On this PR it gives better labeling (and more uniform line widths to match `evoked.plot`:

pr

The interesting thing is that our magnetometer scaling gets a bit far from the +/- 1.96 limits for SSS'ed data. Adding `evoked.plot(noise_cov=cov)` above we get:

scaling

I think is to be expected, though. It actually exists in `plot_white`, too, although it's harder to see because grad and mag are overlayed. I added a note in the docstrings that this is expected and that `plot_white` can be used for additional inspection.

**Comment by larsoner** [comment]
Examples (some didn't change much): [plot_whitened](https://6914-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.0kA2dvG/html/auto_tutorials/plot_whitened.html), [plot_compute_covariance](https://6914-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.0kA2dvG/html/auto_tutorials/plot_compute_covariance.html), [plot_evoked_whitening](https://6914-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.0kA2dvG/html/auto_examples/visualization/plot_evoked_whitening.html), [plot_covariance_whitening_dspm](https://6914-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.0kA2dvG/html/auto_examples/inverse/plot_covariance_whitening_dspm.html).

File: examples/visualization/plot_evoked_whitening.py [hunk 4]
```diff
@@ -5,9 +5,9 @@
 
 Evoked data are loaded and then whitened using a given noise covariance
 matrix. It's an excellent quality check to see if baseline signals match
-the assumption of Gaussian white noise from which we expect values around
-0 with less than 2 standard deviations. Covariance estimation and diagnostic
-plots are based on [1]_.
+the assumption of Gaussian white noise during the baseline period.
```
Comment (dengemann) [inline examples/visualization/plot_evoked_whitening.py]: can we specify +/- 1.96 here as expectation?

Comment (dengemann) [reply examples/visualization/plot_evoked_whitening.py]: Maybe it's enough below.

File: mne/viz/epochs.py [hunk 5]
```diff
@@ -749,7 +748,19 @@ def plot_epochs(epochs, picks=None, scalings=None, n_epochs=20, n_channels=20,
...
 
-        .. versionadded:: 0.15
+        .. versionadded:: 0.15.0
+    noise_cov : instance of Covariance | str | None
+        Noise covariance used to whiten the data while plotting.
+        Whitened data channels are scaled by ``scalings['whitened']``,
+        and their channel names are shown in italic.
+        Can be a string to load a covariance from disk.
+        See also :meth:`mne.Evoked.plot_white` for additional inspection
+        of noise covariance properties when whitening evoked data.
```
Comment (dengemann) [inline mne/viz/epochs.py]: , specifically, for inspection of expected values and model violations ... {or so}

File: mne/viz/evoked.py [hunk 6]
```diff
@@ -579,6 +581,17 @@ def plot_evoked(evoked, picks=None, exclude='bads', unit=True, show=True,
 
         .. versionadded:: 0.13.0
 
+    noise_cov : instance of Covariance | str | None
+        Noise covariance used to whiten the data while plotting.
+        Whitened data channel names are shown in italic.
+        Can be a string to load a covariance from disk.
+        See also :meth:`mne.Evoked.plot_white` for additional inspection
+        of noise covariance properties when whitening evoked data.
+        For data processed with SSS, the effective dependence between
```
Comment (dengemann) [inline mne/viz/evoked.py]: whatever is changed above should also be changed here (model violations / expected values etc).

File: mne/viz/evoked.py [hunk 7]
```diff
@@ -839,26 +863,20 @@ def _plot_update_evoked(params, bools):
...
     Plots the whitened evoked response and the whitened GFP as described in
-    [1]_. If one single covariance object is passed, the GFP panel (bottom)
-    will depict different sensor types. If multiple covariance objects are
-    passed as a list, the left column will display the whitened evoked
-    responses for each channel based on the whitener from the noise covariance
-    that has the highest log-likelihood. The left column will depict the
-    whitened GFPs based on each estimator separately for each sensor type.
-    Instead of numbers of channels the GFP display shows the estimated rank.
-    Note. The rank estimation will be printed by the logger for each noise
-    covariance estimator that is passed.
```
Comment (dengemann) [inline mne/viz/evoked.py]: why is this removed? There is a lot of specific information. I think this is not a good idea.

Comment (dengemann) [reply mne/viz/evoked.py]: Ok. Below. Nevermind.

File: mne/viz/evoked.py [hunk 8]
```diff
@@ -931,102 +944,31 @@ def _plot_evoked_white(evoked, noise_cov, scalings=None, rank=None, show=True):
...
-                    this_estimated_rank, this_picks, this_info, evoked,
-                    cov, ch_type, has_meg, has_sss)
-                this_rank[ch_type] = this_estimated_rank
-            elif rank.get(ch_type) is not None:
-                this_rank[ch_type] = rank[ch_type]
-
-        rank_list.append(this_rank)
+    evoked.pick_types(meg=True, eeg=True, ref_meg=False, exclude='bads')
+    n_ch_used, rank_list, picks_list, has_sss = _triage_rank_sss(
+        evoked.info, noise_cov, rank, scalings)
```
Comment (dengemann) [inline mne/viz/evoked.py]: @larsoner is this call replacing all the work of the helper functions and other lines you removed?

File: mne/viz/ica.py [hunk 9]
```diff
@@ -775,7 +775,8 @@ def _plot_sources_raw(ica, raw, picks, exclude, start, stop, show, title,
                   ch_start=0, t_start=start, info=info, duration=duration,
                   ica=ica, n_channels=n_channels, times=times, types=types,
                   n_times=raw.n_times, bad_color=bad_color, picks=picks,
-                  first_time=first_time, data_picks=[], decim=1)
+                  first_time=first_time, data_picks=[], decim=1,
+                  noise_cov=None, whitened_ch_names=())
```
Comment (dengemann) [inline mne/viz/ica.py]: @larsoner why are we doing this actually? Just because we can or because people will need this. How does our whitening interact with ICA?

File: mne/viz/utils.py [hunk 10]
```diff
@@ -2242,3 +2250,192 @@ def _grad_pair_pick_and_name(info, picks):
...
+        assert set(whitened_ch_names) == whiten_names
+        projector[whiten_picks, whiten_picks[:, np.newaxis]] = whitener
+        # Now we need to change the set of "whitened" channels to include
+        # all data channel names so that they are properly italicized.
+        whitened_ch_names = data_names
+        # We would need to set "bad_picks" to identity to show the traces
+        # (but in gray), but here we don't need to because "projector"
+        # starts out as identity. So all that is left to do is take any
+        # *good* data channels that are not in the noise cov to be NaN
+        nan_names = data_names - (bad_names | cov_names)
```
Comment (dengemann) [inline mne/viz/utils.py]: maybe the case for using set methods for better readability.

**Comment by dengemann** [comment]
@larsoner I am not sure it is the best example. Can we plot the whitening of the evoked to see what is going on? As we discussed, the magnetometers are expected to be much lower in amplitude after SSS. Together with gradiometers they make up n_sss_components dimensions, as is shown in the GFP plot of the whitening.
Also to be on the same page, can you describe what was that bug that you fixed? / point to the buggy code you replaced with correct code?

We're getting somewhere, don't give up :) I think we're moving in the right direction .

File: mne/viz/utils.py (hunk 10 above)
Comment (larsoner) [reply mne/viz/utils.py]: I think set operations are clearly described using their arithmetic equivalents ...?

File: mne/viz/ica.py (hunk 9 above)
Comment (larsoner) [reply mne/viz/ica.py]: Need to pass the params / API uniformity internally. They shouldn't use it with ICA which is why it's not exposed in the public facing APIs for ICA plotting.

File: mne/viz/evoked.py (hunk 8 above)
Comment (larsoner) [reply mne/viz/evoked.py]: Yes it was copy-paste to refactor + bug fixes

**Comment by larsoner** [comment]
> @larsoner I am not sure it is the best example. Can we plot the whitening of the evoked to see what is going on?

Which example? The script I posted above? The three images in my previous post all come from the same script so it has `plot_white` as well as `evoked.plot(noise_cov=cov)` outputs. The traces are the same you just can't see the mags in the `plot_white` case because they are small relative to the grads. (I checked this by manually `.set_visible(False)`ing the gradiometer lines in the `plot_white` case.)

File: mne/viz/epochs.py (hunk 5 above)
Comment (larsoner) [reply mne/viz/epochs.py]: I'm wary of adding more text to an already long description here. I'll add it to the text for `plot_evoked_white`

File: mne/viz/utils.py [hunk 11]
```diff
@@ -2242,3 +2250,192 @@ def _grad_pair_pick_and_name(info, picks):
...
+        picks_list2 += _picks_by_type(info, meg_combined=True)
+
+    rank_list = []  # rank dict for each cov
+    for cov in covs:
+        # We need to add the covariance projectors, compute the projector,
+        # and apply it, just like we will do in prepare_noise_cov, otherwise
+        # we risk the rank estimates being incorrect (i.e., if the projectors
+        # do not match).
+        info_proj = info.copy()
+        info_proj['projs'] += cov['projs']
```
Comment (larsoner) [inline mne/viz/utils.py]: This is one bugfix -- the estimated rank needs to use the union of projectors to mirror what happens in computing the whitener

File: mne/viz/utils.py [hunk 12]
```diff
@@ -2242,3 +2250,192 @@ def _grad_pair_pick_and_name(info, picks):
...
+                # and we're done here
+                break
+
+            if rank.get(ch_type) is None:
+                this_info = pick_info(info_proj, this_picks)
+                idx = np.ix_(this_picks, this_picks)
+                projector = setup_proj(this_info, add_eeg_ref=False)[0]
+                this_C = C[idx]
+                if projector is not None:
+                    this_C = np.dot(np.dot(projector, this_C), projector.T)
```
Comment (larsoner) [inline mne/viz/utils.py]: ... and correspondingly this joint projector actually needs to be applied to the covariance matrix to ensure rank estimation works properly

File: mne/viz/evoked.py [hunk 13]
```diff
@@ -1081,9 +1023,12 @@ def whitened_gfp(x, rank=None):
...
     else:
         for ((ch_type, picks), ax) in zip(picks_list, axes_evoked):
-            ax.plot(times, evokeds_white[0].data[picks].T, color='k')
+            ax.plot(times, evokeds_white[0].data[picks].T, color='k',
+                    lw=0.5)
             for hline in [-1.96, 1.96]:
-                ax.axhline(hline, color='red', linestyle='--')
+                ax.axhline(hline, color='red', linestyle='--', lw=2)
+            ax.set(title='{0} ({1} channels)'.format(titles_[ch_type],
+                                                     len(picks)))
```
Comment (larsoner) [inline mne/viz/evoked.py]: this is a bug / consistency fix (the one I show with a script in a previous comment)

**Comment by larsoner** [comment]
@dengemann comments addressed (I think)

**Comment by larsoner** [comment]
ping @dengemann let me know if this is good to go for you

**Comment by dengemann** [comment]
Looks good. We should move on. I let @agramfort take another look.

**Comment by larsoner** [comment]
@agramfort okay for you too?
